File: src/accountInfo.ts
/**
 * Account + plan awareness (TRDD-BURNWDGT). Resolves the CURRENTLY-active Claude Code OAuth account
 * (identity + plan type) so the per-account window budget can label its buckets and answer "which
 * account am I on, what plan, how much of its window is left".
 *
 * TWO live-only sources (never persisted — this is display metadata that drifts, and one source holds
 * a secret):
 *   1. ~/.claude.json  `oauthAccount`  → accountUuid, emailAddress, organizationName/Uuid, billingType,
 *      hasExtraUsageEnabled, rate-limit tiers, displayName. These are IDENTITY fields, no secret.
 *   2. macOS keychain `Claude Code-credentials` → the credential JSON carries the OAuth TOKENS **and**
 *      `subscriptionType` (max / pro / team / enterprise / free). We read the blob, extract ONLY
 *      `subscriptionType`, and drop the rest. The access/refresh TOKEN is NEVER returned, logged, or
 *      persisted — parseSubscriptionType is the single choke-point and returns a plain plan string.
 *
 * Everything is fail-soft: a missing/malformed file, a denied keychain read, or a non-macOS host yields
 * nulls, never a throw and never a secret. The keychain read is injectable so it is unit-testable
 * without touching the real keychain (and defaults to a short-timeout, non-shell execFile).
 */

import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import { execFileSync } from 'child_process'

/** The identity fields lifted from ~/.claude.json `oauthAccount` (no secrets live here). */
export interface OauthIdentity {
  accountUuid: string | null
  email: string | null
  organizationName: string | null
  organizationUuid: string | null
  billingType: string | null           // e.g. "subscription" | "api" — subscription = window-limited
  hasExtraUsageEnabled: boolean         // true = opted into token billing once the window is hit
  organizationRateLimitTier: string | null
  userRateLimitTier: string | null
  displayName: string | null
}

/** The fully-resolved, live-only account view surfaced to the MCP / budget layer. */
export interface AccountInfo extends OauthIdentity {
  planType: string | null              // keychain subscriptionType (max/pro/team/enterprise/free)
  rateLimitTier: string | null         // organizationRateLimitTier ?? userRateLimitTier
  label: string                        // human label: email ?? displayName ?? short accountUuid
  source: 'claude.json' | 'none'
}

function str(v: unknown): string | null {
  return typeof v === 'string' && v.length > 0 ? v : null
}

/** Parse ~/.claude.json's `oauthAccount` into the identity fields. Pure + fail-soft: returns null on any
 *  malformed input so a corrupt config never crashes the monitor. No secret is present in this object. */
export function parseOauthAccount(claudeJsonText: string): OauthIdentity | null {
  let oa: Record<string, unknown>
  try {
    const root = JSON.parse(claudeJsonText) as Record<string, unknown>
    const acct = root.oauthAccount
    if (!acct || typeof acct !== 'object') { return null }
    oa = acct as Record<string, unknown>
  } catch {
    return null
  }
  return {
    accountUuid: str(oa.accountUuid),
    email: str(oa.emailAddress),
    organizationName: str(oa.organizationName),
    organizationUuid: str(oa.organizationUuid),
    billingType: str(oa.billingType),
    hasExtraUsageEnabled: oa.hasExtraUsageEnabled === true,
    organizationRateLimitTier: str(oa.organizationRateLimitTier),
    userRateLimitTier: str(oa.userRateLimitTier),
    displayName: str(oa.displayName),
  }
}

/** Extract ONLY the plan type (`subscriptionType`) from the keychain credential JSON. THE SECURITY
 *  CHOKE-POINT: the credential blob also holds the OAuth access/refresh tokens; this function reads a
 *  single non-secret string and returns it — it never returns, logs, or exposes any token field.
 *  Fail-soft: null on any parse failure or a missing subscriptionType. */
export function parseSubscriptionType(credentialJsonText: string): string | null {
  try {
    const o = JSON.parse(credentialJsonText) as Record<string, unknown>
    // Claude Code stores the OAuth blob under `claudeAiOauth`; older shapes put it at the top level.
    const inner = (o.claudeAiOauth && typeof o.claudeAiOauth === 'object')
      ? o.claudeAiOauth as Record<string, unknown>
      : o
    return str(inner.subscriptionType)
  } catch {
    return null
  }
}

/** One-shot latch for the keychain plan read. WHY THIS EXISTS (bug autopsy, 2026-07-09): reading the
 *  `Claude Code-credentials` item triggers a macOS "Security wants to use the login keychain" password
 *  dialog whenever the item's ACL does NOT include our binary — which happens after EVERY account
 *  rotation, because Claude Code re-creates the item with a Claude-Code-only ACL that excludes node.
 *  A per-tick reader (burn tick / per-request / per-respawn) then FLOODS the user with hundreds of
 *  un-dismissable prompts (macOS offers no "Always Allow" because the reader isn't in the ACL). The
 *  60s getCurrentAccount cache did not prevent it (opts-callers bypass it, and each supervisor respawn
 *  starts a cold process that prompts on its first tick). So the plan-from-keychain read is now
 *  STRICTLY OPT-IN and attempted AT MOST ONCE per process: any outcome (success, denial, timeout,
 *  off-macOS, not-opted-in) latches permanently so it can NEVER re-prompt. The plan label is
 *  display-only; identity + billingType + rate-limit tiers already come from ~/.claude.json with no
 *  prompt, so losing the keychain plan degrades to null, never to a flood. */
let keychainPlanLatch: { done: boolean; value: string | null } = { done: false, value: null }

/** Read the keychain credential blob on macOS and return ONLY the subscriptionType — but ONLY when the
 *  user explicitly opts in via AGENTLENS_READ_KEYCHAIN_PLAN=1, and AT MOST ONCE per process (latched).
 *  Off-macOS, not-opted-in, denied, or any error → null, latched, never retried. The raw blob (which
 *  holds the OAuth tokens) is confined to this function's local scope and never surfaced. */
function readKeychainSubscriptionType(): string | null {
  if (keychainPlanLatch.done) { return keychainPlanLatch.value }
  // Default OFF: never touch the keychain unless explicitly opted in — a single un-ACL'd read prompts.
  if (process.platform !== 'darwin' || process.env.AGENTLENS_READ_KEYCHAIN_PLAN !== '1') {
    keychainPlanLatch = { done: true, value: null }
    return null
  }
  let value: string | null = null
  try {
    const raw = execFileSync('security',
      ['find-generic-password', '-s', 'Claude Code-credentials', '-w'],
      { encoding: 'utf8', timeout: 3000, stdio: ['ignore', 'pipe', 'ignore'] })
    value = parseSubscriptionType(raw)
  } catch {
    value = null   // denial / timeout / errSecAuthFailed — latch OFF, never re-prompt this process
  }
  keychainPlanLatch = { done: true, value }
  return value
}

/** A human label for the account: email, else org display name, else a short account_uuid, else 'unknown'. */
export function accountLabelFor(id: OauthIdentity | null, accountUuid?: string | null): string {
  if (id && (accountUuid == null || id.accountUuid === accountUuid)) {
    return id.email || id.displayName || (id.accountUuid ? id.accountUuid.slice(0, 8) : 'unknown')
  }
  // A different (rotated-away) account is not resolvable to an email — only its id is known.
  return accountUuid ? accountUuid.slice(0, 8) : 'unknown'
}

let cached: { info: AccountInfo; ts: number } | null = null
const ACCOUNT_TTL_MS = 60_000   // avoid re-reading claude.json + the keychain on every burn tick

/** The current live account (identity + plan), cached ~60s to avoid repeated keychain reads. `homeDir`
 *  + `readKeychain` are injectable for tests (so the real keychain is never touched under test). */
export function getCurrentAccount(opts?: {
  homeDir?: string
  readKeychain?: () => string | null
  now?: number
}): AccountInfo {
  const now = opts?.now ?? Date.now()
  if (!opts && cached && now - cached.ts < ACCOUNT_TTL_MS) { return cached.info }

  const home = opts?.homeDir ?? os.homedir()
  let identity: OauthIdentity | null = null
  try {
    identity = parseOauthAccount(fs.readFileSync(path.join(home, '.claude.json'), 'utf8'))
  } catch {
    identity = null
  }
  const planType = (opts?.readKeychain ?? readKeychainSubscriptionType)()

  const info: AccountInfo = identity
    ? {
        ...identity,
        planType,
        rateLimitTier: identity.organizationRateLimitTier ?? identity.userRateLimitTier,
        label: accountLabelFor(identity),
        source: 'claude.json',
      }
    : {
        accountUuid: null, email: null, organizationName: null, organizationUuid: null,
        billingType: null, hasExtraUsageEnabled: false,
        organizationRateLimitTier: null, userRateLimitTier: null, displayName: null,
        planType, rateLimitTier: null, label: 'unknown', source: 'none',
      }

  if (!opts) { cached = { info, ts: now } }
  return info
}

File: rust-core/crates/agentlens-core/src/burn/account_info.rs
//! Port of src/accountInfo.ts (TRDD-BURNWDGT) — the CURRENTLY-active Claude Code OAuth
//! account (identity + plan) for the per-account window labels and "who am I / what plan".
//!
//! Two live-only sources, exactly as in TS: `~/.claude.json` `oauthAccount` (identity fields,
//! no secret) and — STRICTLY OPT-IN (`AGENTLENS_READ_KEYCHAIN_PLAN=1`), macOS-only, latched to
//! AT MOST ONCE per process — the keychain credential's `subscriptionType`. The credential
//! blob also holds the OAuth tokens; `parse_subscription_type` is the single choke-point and
//! returns only the plan string. Everything is fail-soft: nulls, never a throw, never a secret.

use serde_json::{Map, Value};
use std::path::Path;
use std::sync::OnceLock;

use crate::summarize::helpers::js_slice;

#[derive(Clone, Debug, Default)]
pub struct AccountInfo {
    pub account_uuid: Option<String>,
    pub email: Option<String>,
    pub organization_name: Option<String>,
    pub organization_uuid: Option<String>,
    /// e.g. "stripe_subscription" | "api" — any value CONTAINING 'subscription' is window-limited.
    pub billing_type: Option<String>,
    pub has_extra_usage_enabled: bool,
    pub organization_rate_limit_tier: Option<String>,
    pub user_rate_limit_tier: Option<String>,
    pub display_name: Option<String>,
    pub plan_type: Option<String>,
    pub rate_limit_tier: Option<String>,
    pub label: String,
    pub source: &'static str, // "claude.json" | "none"
}

impl AccountInfo {
    /// The TS AccountInfo object shape (identity spread first, then the four derived fields).
    pub fn to_value(&self) -> Value {
        let os = |v: &Option<String>| v.clone().map_or(Value::Null, Value::from);
        let mut m = Map::new();
        m.insert("accountUuid".into(), os(&self.account_uuid));
        m.insert("email".into(), os(&self.email));
        m.insert("organizationName".into(), os(&self.organization_name));
        m.insert("organizationUuid".into(), os(&self.organization_uuid));
        m.insert("billingType".into(), os(&self.billing_type));
        m.insert("hasExtraUsageEnabled".into(), Value::Bool(self.has_extra_usage_enabled));
        m.insert("organizationRateLimitTier".into(), os(&self.organization_rate_limit_tier));
        m.insert("userRateLimitTier".into(), os(&self.user_rate_limit_tier));
        m.insert("displayName".into(), os(&self.display_name));
        m.insert("planType".into(), os(&self.plan_type));
        m.insert("rateLimitTier".into(), os(&self.rate_limit_tier));
        m.insert("label".into(), self.label.clone().into());
        m.insert("source".into(), self.source.into());
        Value::Object(m)
    }
}

/// TS `str()` — a non-empty string, else null.
fn nes(v: Option<&Value>) -> Option<String> {
    v.and_then(Value::as_str).filter(|s| !s.is_empty()).map(str::to_owned)
}

/// parseOauthAccount — the identity half of AccountInfo from ~/.claude.json text. None on any
/// malformed input (a corrupt config never crashes the monitor). The derived fields (plan,
/// tier, label, source) are filled by `get_current_account`.
pub fn parse_oauth_account(claude_json_text: &str) -> Option<AccountInfo> {
    let root: Value = serde_json::from_str(claude_json_text).ok()?;
    let oa = root.get("oauthAccount").filter(|v| v.is_object())?;
    Some(AccountInfo {
        account_uuid: nes(oa.get("accountUuid")),
        email: nes(oa.get("emailAddress")),
        organization_name: nes(oa.get("organizationName")),
        organization_uuid: nes(oa.get("organizationUuid")),
        billing_type: nes(oa.get("billingType")),
        has_extra_usage_enabled: oa.get("hasExtraUsageEnabled") == Some(&Value::Bool(true)),
        organization_rate_limit_tier: nes(oa.get("organizationRateLimitTier")),
        user_rate_limit_tier: nes(oa.get("userRateLimitTier")),
        display_name: nes(oa.get("displayName")),
        ..AccountInfo::default()
    })
}

/// parseSubscriptionType — THE SECURITY CHOKE-POINT: from the keychain credential JSON (which
/// also holds the OAuth tokens) extract ONLY the plan string; never any token field.
pub fn parse_subscription_type(credential_json_text: &str) -> Option<String> {
    let o: Value = serde_json::from_str(credential_json_text).ok()?;
    let inner = o.get("claudeAiOauth").filter(|v| v.is_object()).unwrap_or(&o);
    nes(inner.get("subscriptionType"))
}

/// accountLabelFor — email, else display name, else short uuid, else 'unknown'; a DIFFERENT
/// (rotated-away) uuid resolves only to its short id.
pub fn account_label_for(id: Option<&AccountInfo>, account_uuid: Option<&str>) -> String {
    if let Some(id) = id {
        if account_uuid.is_none() || id.account_uuid.as_deref() == account_uuid {
            return id
                .email
                .clone()
                .or_else(|| id.display_name.clone())
                .or_else(|| id.account_uuid.as_deref().map(|u| js_slice(u, 8).to_owned()))
                .unwrap_or_else(|| "unknown".to_owned());
        }
    }
    account_uuid.map_or_else(|| "unknown".to_owned(), |u| js_slice(u, 8).to_owned())
}

/// The one-shot keychain latch (bug autopsy 2026-07-09, kept verbatim from TS): an un-ACL'd
/// read PROMPTS the user, so the read is opt-in AND attempted at most once per process —
/// any outcome latches. Process-global deliberately: the safety property must hold across
/// every BurnRuntime a test constructs.
static KEYCHAIN_PLAN: OnceLock<Option<String>> = OnceLock::new();

fn read_keychain_subscription_type(opted_in: bool) -> Option<String> {
    KEYCHAIN_PLAN
        .get_or_init(|| {
            if !cfg!(target_os = "macos") || !opted_in {
                return None;
            }
            let mut child = std::process::Command::new("security")
                .args(["find-generic-password", "-s", "Claude Code-credentials", "-w"])
                .stdin(std::process::Stdio::null())
                .stdout(std::process::Stdio::piped())
                .stderr(std::process::Stdio::null())
                .spawn()
                .ok()?;
            // The TS uses a 3s execFile timeout; poll-wait the same ceiling, then kill.
            let deadline = std::time::Instant::now() + std::time::Duration::from_secs(3);
            loop {
                match child.try_wait() {
                    Ok(Some(status)) => {
                        if !status.success() {
                            return None;
                        }
                        let mut out = String::new();
                        use std::io::Read;
                        child.stdout.take()?.read_to_string(&mut out).ok()?;
                        return parse_subscription_type(&out);
                    }
                    Ok(None) => {
                        if std::time::Instant::now() >= deadline {
                            let _ = child.kill();
                            return None;
                        }
                        std::thread::sleep(std::time::Duration::from_millis(50));
                    }
                    Err(_) => return None,
                }
            }
        })
        .clone()
}

/// getCurrentAccount (uncached — the 60s cache lives on BurnRuntime). `read_keychain` is the
/// injectable plan resolver, as in TS; None = the real latched keychain read.
pub fn get_current_account(home_dir: &Path, vars: &std::collections::HashMap<String, String>, read_keychain: Option<&dyn Fn() -> Option<String>>) -> AccountInfo {
    let identity = std::fs::read_to_string(home_dir.join(".claude.json")).ok().and_then(|t| parse_oauth_account(&t));
    let plan_type = match read_keychain {
        Some(f) => f(),
        None => read_keychain_subscription_type(vars.get("AGENTLENS_READ_KEYCHAIN_PLAN").map(String::as_str) == Some("1")),
    };
    match identity {
        Some(id) => {
            let label = account_label_for(Some(&id), None);
            AccountInfo {
                plan_type,
                rate_limit_tier: id.organization_rate_limit_tier.clone().or_else(|| id.user_rate_limit_tier.clone()),
                label,
                source: "claude.json",
                ..id
            }
        }
        None => AccountInfo { plan_type, label: "unknown".to_owned(), source: "none", ..AccountInfo::default() },
    }
}
